devtools/common/src/main/java/io/quarkus/cli/commands/AddExtensions.java

```diff
@@ -31,57 +33,96 @@ public AddExtensions(final ProjectWriter writer, final String pom) throws IOExce
+                artifactId.equalsIgnoreCase("quarkus-" + q) ||
+                artifactId.equalsIgnoreCase("quarkus-smallrye-" + q));
+    }
+
+    public AddExtensionResult addExtensions(final Set<String> extensions) throws IOException {
         if (extensions == null || extensions.isEmpty()) {
-            return false;
+            return new AddExtensionResult(false, true);
         }
 
         boolean updated = false;
+        boolean success = true;
         List<Dependency> dependenciesFromBom = getDependenciesFromBom();
 
-        for (String dependency : extensions) {
-            List<Extension> matches = MojoUtils.loadExtensions().stream()
-                    .filter(d -> {
-                        boolean hasTag = d.labels().contains(dependency.trim().toLowerCase());
-                        boolean machName = d.getName()
-                                .toLowerCase()
-                                .contains(dependency.trim().toLowerCase());
-                        boolean matchArtifactId = d.getArtifactId()
-                                .toLowerCase()
-                                .contains(dependency.trim().toLowerCase());
-                        return hasTag || machName || matchArtifactId;
-                    })
-                    .collect(Collectors.toList());
-
-            if (matches.size() > 1) {
+        for (String query : extensions) {
+            List<Extension> registry = MojoUtils.loadExtensions();
+
+            SelectionResult result = select(query, registry);
+
+            if (!result.matches()) {
                 StringBuilder sb = new StringBuilder();
-                sb.append(NOK)
-                        .append(" Multiple extensions matching '" + dependency + "'");
-
-                matches.stream()
-                        .forEach(extension -> sb.append(System.lineSeparator()).append("     * ")
-                                .append(extension.managementKey()));
-                sb.append(System.lineSeparator()).append("     Be more specific e.g using the exact name or the full gav.");
-                System.out.println(sb);
-            } else if (matches.size() == 1) {
-                final Extension extension = matches.get(0);
-
-                if (!MojoUtils.hasDependency(model, extension.getGroupId(), extension.getArtifactId())) {
-                    System.out.println(OK + " Adding extension " + extension.managementKey());
-                    model.addDependency(extension
-                            .toDependency(containsBOM(model) &&
-                                    isDefinedInBom(dependenciesFromBom, extension)));
-                    updated = true;
+                // We have 3 cases, we can still have a single candidate, but the match is on label
+                // or we have several candidates, or none
+                Set<Extension> candidates = result.getExtensions();
+                if (candidates.isEmpty() && query.contains(":")) {
```

Comment:
`query.contains(":")` always beats candidates I think. It's an absolute name.

Reply:
Agreed.

Reply:
in addition, the insertion must avoid duplicates, so 2 bugs.

Reply:
Fixed